File: binaryThree/binaryThree.go
package main
import "fmt"

type Node struct {
  value int
  parent *Node
  next *Node
  previous *Node
}

type BinaryThree struct {
  head Node
}

func newBinaryThree(value int) *BinaryThree {
  n := &Node{value: value}
  return &BinaryThree{*n}
}

func (bt *BinaryThree) add(value int) *Node {
  temp := &bt.head
  for true {
    if temp.value == value {
      break
    }

    if temp.value < value {
      if temp.next == nil {
        temp.next = &Node{value: value, parent: temp} 
        temp = temp.next
        break
      }
      temp = temp.next
    }

    if temp.value > value {
      if temp.previous == nil {
        temp.previous = &Node{value: value, parent: temp} 
        temp = temp.previous
        break
      }
      temp = temp.previous
    }
  }

  return temp
}

func main() {
  bt := BinaryThree{}
  bt.add(10)
  fmt.Println(bt.head.next.value)
}

File: binaryThree/binaryThree.ts
interface Element {
  value: number;
  parent?: Element;
  next?: Element;
  previous?: Element;
}

class binaryThree {
  public head: Element;

  constructor(value?: number) {
    this.head = value ? { value } : null!;
  }

  public add(value: number): Element {
    if (!this.head) {
      this.head = { value };
      return this.head;
    }

    let temp = this.head;

    while (true) {
      if (temp.value === value) {
        return temp;
      }

      if (temp.value < value) {
        if (!temp.next) {
          temp.next = {
            value,
            parent: temp
          }
          return temp.next;
        }
        temp = temp.next;
      }

      if (temp.value > value) {
        if (!temp.previous) {
          temp.previous = {
            value,
            parent: temp
          }
          return temp.previous;
        }
        temp = temp.previous;
      }
    }
  }
}

const bt = new binaryThree(4);

for (let count = 0; count < 10; count++) {
  let randomNumer = Math.floor(Math.random() * 50) + 1;
  bt.add(randomNumer);
}

let lessValue = bt.head;
while (true) {
  console.log(lessValue);
  if (!lessValue.next) break;
  lessValue = lessValue.next;
}
